core/payments/ondemand/cumulative_payment_store.go

```diff
@@ -2,26 +2,130 @@ package ondemand
-	// Returns an [ondemand.InsufficientFundsError] if the addition would cause the cumulative payment to exceed
-	// the input maxCumulativePayment. In this case, the underlying cumulative payment value is not modified.
-	AddCumulativePayment(
-		ctx context.Context,
-		// the amount to add to the cumulative payment, in wei
-		amount *big.Int,
-		// the maximum value for the cumulative payment, after the addition. Should be determined by the `TotalDeposits`
-		// for the account in the payment vault
-		maxCumulativePayment *big.Int,
-	) (*big.Int, error)
+// TODO(litt3): there are some potential avenues for optimization of this store:
+// 1. Use something other than DynamoDB. DynamoDB is being used for historical reasons, but there is only a single
+// writer now, which doesn't need any of the distributed DB properties provided by DynamoDB.
+// 2. Implement a write queue, so that the caller doesn't need to wait for the write to complete. The callers of the
+// CumulativePaymentStore just need *eventual* persistence of the cumulative payment, so using a queue would be
+// sufficient, and would free the caller from blocking on I/O.
+type CumulativePaymentStore struct {
+	// The DynamoDB client to use for storage operations
+	dynamoClient *dynamodb.Client
+	// The name of the DynamoDB table to store payments in, stored as *string for use in DynamoDB operations
+	tableName *string
+	// The account address, pre-built as a key for DynamoDB operations
+	accountKey map[string]types.AttributeValue
+}
+
+// Creates a new DynamoDB-backed cumulative payment store
+func NewCumulativePaymentStore(
+	dynamoClient *dynamodb.Client,
+	tableName string,
+	// The account ID this store is tracking payments for
+	accountID gethcommon.Address,
+) (*CumulativePaymentStore, error) {
+	if dynamoClient == nil {
+		return nil, fmt.Errorf("dynamoClient cannot be nil")
+	}
+	if tableName == "" {
+		return nil, fmt.Errorf("tableName cannot be empty")
+	}
+	if accountID == (gethcommon.Address{}) {
+		return nil, fmt.Errorf("accountID cannot be the zero address")
+	}
+
+	return &CumulativePaymentStore{
+		dynamoClient: dynamoClient,
+		tableName:    aws.String(tableName),
+		accountKey: map[string]types.AttributeValue{
+			attributeAccountID: &types.AttributeValueMemberS{Value: accountID.Hex()},
+		},
+	}, nil
+}
+
+// Stores a new cumulative payment value in DynamoDB
+func (s *CumulativePaymentStore) StoreCumulativePayment(
+	ctx context.Context,
+	newCumulativePayment *big.Int,
+) error {
+	if newCumulativePayment == nil {
+		return errors.New("newCumulativePayment cannot be nil")
+	}
+	if newCumulativePayment.Sign() < 0 {
+		return fmt.Errorf("cumulative payment cannot be negative: received %s", newCumulativePayment.String())
+	}
+
+	_, err := s.dynamoClient.UpdateItem(ctx, &dynamodb.UpdateItemInput{
+		TableName:        s.tableName,
+		Key:              s.accountKey,
+		UpdateExpression: aws.String("SET #cp = :new"),
```

Comment:
```suggestion
		UpdateExpression: aws.String("SET #cp = :newValue"),
```

Reply:
was confused for a little while until I realized `:new` was just a variable name

Reply:
e6b4f2c2df758e4268dfcfea048eb5f531cad72e